x-pack/spec/license_checker/license_reader_spec.rb

```diff
@@ -51,22 +51,40 @@
   describe '#fetch_xpack_info' do
     let(:xpack_info_class) { LogStash::LicenseChecker::XPackInfo }
     let(:mock_client) { double('Client') }
-    before(:each) { expect(subject).to receive(:client).and_return(mock_client) }
+    before(:each) { expect(subject).to receive(:client).and_return(mock_client).at_most(:twice) }
+    let(:xpack_info) do
+      {
+          "license" => {},
+          "features" => {},
+      }
+    end
 
     context 'when client fetches xpack info' do
-      let(:xpack_info) do
-        {
-            "license" => {},
-            "features" => {},
-        }
-      end
       before(:each) do
         expect(mock_client).to receive(:get).with('_xpack').and_return(xpack_info)
       end
       it 'returns an XPackInfo' do
         expect(subject.fetch_xpack_info).to eq(xpack_info_class.from_es_response(xpack_info))
       end
     end
+
+    context 'and receives HostUnreachableError' do
+      let(:host_not_reachable) { LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError.new(StandardError.new("original error"), "http://localhost:19200") }
+      before(:each) do
+        first_call = true
+        allow(mock_client).to receive(:get) do |path|
+          expect(path).to eq('_xpack')
+          if first_call
+            first_call = false
+            raise host_not_reachable
+          end
+          xpack_info
+        end
+      end
```

Comment:
We can rely on multiple ordered expectations on our partial mock, which should make it more readable.

```suggestion
      before(:each) do
        # set up expectation of single failure
        expect(license_reader.logger).to receive(:warn).with(a_string_including "Attempt to validate Elasticsearch license failed.")
        expect(mock_client).to receive(:get).with('_xpack').and_raise(host_not_reachable).once
        
        # ensure subsequent success
        expect(mock_client).to receive(:get).with('_xpack').and_return(xpack_info)
      end
```

Reply:
You are right, far better, I missed I could instruct multiple responses for same call.